### PR TITLE
Replace most `makeChainedTransactions` with `TxBuilder`

### PR DESCRIPTION
**Pull request by kchulj**
`makeChainedTransactions` is an outdated function which assumes a block always contains exactly 8 transactions.
Since we have transitioned away from creating a block based on the number of transactions, we should use `TxBuilder` instead.

Relates #1107

### CODE REVIEW

File: source/agora/utils/Test.d [hunk 1]
```diff
@@ -736,13 +736,17 @@ public struct TxBuilder
 }
```
Comment (AndrejMitrovic) [inline source/agora/utils/Test.d]: This commit should be squashed with the previous one.

File: source/agora/common/crypto/Key.d [hunk 2]
```diff
@@ -120,8 +120,9 @@ public struct PublicKey
     }
 
     /// Uses Stellar's representation instead of hex
-    public string toString () const @trusted
+    public string toString () const @trusted nothrow
     {
+        scope (failure) assert(0);
```
Comment (AndrejMitrovic) [inline source/agora/common/crypto/Key.d]: Unrelated change. I think you may want to rebase your PR first because this was merged a little while ago: https://github.com/bpfkorea/agora/pull/1125

File: source/agora/consensus/data/Block.d [hunk 3]
```diff
@@ -439,7 +439,7 @@ unittest
 *******************************************************************************/
 
 public Block makeNewBlock (Transactions)(const ref Block prev_block,
-    Transactions txs, Enrollment[] enrollments) @safe nothrow
+    Transactions txs, Enrollment[] enrollments) @safe
```
Comment (AndrejMitrovic) [inline source/agora/consensus/data/Block.d]: Ditto comment about https://github.com/bpfkorea/agora/pull/1125 being merged, so I think this should remain being `nothrow`.

File: source/agora/consensus/validation/Block.d [hunk 4]
```diff
@@ -435,7 +435,7 @@ unittest
     auto gen_hash = GenesisBlock.header.hashFull();
 
     GenesisBlock.txs.each!(tx => utxos.put(tx));
-    auto block = GenesisBlock.makeNewBlock(makeChainedTransactions(gen_key, null, 1));
+    auto block = GenesisBlock.makeNewBlock(genesisSpendable().map!(txb => txb.sign()).array());
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
    auto block = GenesisBlock.makeNewBlock(genesisSpendable().map!(txb => txb.sign()));
```

`makeNewBlock` accepts an input range, so there's no need to call `.array` - it's an unnecessary allocation.

File: source/agora/consensus/validation/Block.d [hunk 5]
```diff
@@ -498,7 +498,7 @@ unittest
     prev_txs.each!(tx => utxos.put(tx));  // these will be spent
 
     auto prev_block = block;
-    block = block.makeNewBlock(makeChainedTransactions(gen_key, prev_txs, 1));
+    block = block.makeNewBlock(prev_txs.map!(tx => TxBuilder(tx).sign()).array());
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
    block = block.makeNewBlock(prev_txs.map!(tx => TxBuilder(tx).sign()));
```

File: source/agora/consensus/validation/Block.d [hunk 6]
```diff
@@ -547,7 +547,7 @@ unittest
     };
 
     // consumed all utxo => fail
-    block = GenesisBlock.makeNewBlock(makeChainedTransactions(gen_key, null, 1));
+    block = GenesisBlock.makeNewBlock(genesisSpendable().map!(txb => txb.sign()).array());
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
    block = GenesisBlock.makeNewBlock(genesisSpendable().map!(txb => txb.sign()));
```

File: source/agora/consensus/validation/Block.d [hunk 7]
```diff
@@ -568,7 +568,7 @@ unittest
     // we stopped validation due to a double-spend
     assert(used_set.length == double_spend.length - 1);
 
-    block = GenesisBlock.makeNewBlock(makeChainedTransactions(gen_key, prev_txs, 1));
+    block = GenesisBlock.makeNewBlock(prev_txs.map!(tx => TxBuilder(tx).sign()).array());
```
Comment (AndrejMitrovic) [inline source/agora/consensus/validation/Block.d]:
```suggestion
    block = GenesisBlock.makeNewBlock(prev_txs.map!(tx => TxBuilder(tx).sign()));
```

File: source/agora/test/NetworkManager.d [hunk 8]
```diff
@@ -76,16 +77,22 @@ unittest
 
             Block last_block;
             // make 20 blocks which have an invalid previous hash
-            foreach (idx; 0 .. 20)
+
+            // genesis case
+            if (!last_tx.length)
```
Comment (AndrejMitrovic) [inline source/agora/test/NetworkManager.d]: Styling nit: when there's only a single statement then there is no need to add braces. So:

```D
if (!last_tx.length)
    last_tx = ...
```

File: source/agora/test/NetworkManager.d [hunk 9]
```diff
@@ -76,16 +77,22 @@ unittest
 
             Block last_block;
             // make 20 blocks which have an invalid previous hash
-            foreach (idx; 0 .. 20)
+
```
Comment (AndrejMitrovic) [inline source/agora/test/NetworkManager.d]: I am not sure if this is the correct change. This code segment created invalid transactions **on purpose**, to test the behavior of the system. I will have to double-check the code again tomorrow to see if it does the right thing here.

File: source/agora/test/NetworkManager.d [hunk 10]
```diff
@@ -151,11 +158,20 @@ unittest
 
     Transaction[][] block_txes; /// per-block array of transactions (genesis not included)
     Transaction[] last_txs;
-    foreach (block_idx; 0 .. 10)  // create 10 blocks
+
+    // create genesis block
+    if (!last_txs.length)
```
Comment (AndrejMitrovic) [inline source/agora/test/NetworkManager.d]: This check shouldn't be necessary, `last_txs` is empty here for sure.

File: source/agora/test/NetworkManager.d [hunk 11]
```diff
@@ -151,11 +158,20 @@ unittest
...
     {
-        // create enough tx's for a single block
-        auto txs = makeChainedTransactions(WK.Keys.Genesis, last_txs, 1);
+        last_txs = genesisSpendable().map!(txb => txb.sign()).array();
+        last_txs.each!(tx => node_validator.putTransaction(tx));
+        network.expectBlock([node_validator], Height(1), 4.seconds);
+        block_txes ~= last_txs.sort.array;
+    }
 
+    foreach (block_idx; 1 .. 10)  // create a total of 10 blocks
```
Comment (AndrejMitrovic) [inline source/agora/test/NetworkManager.d]:
```suggestion
    foreach (block_idx; 1 .. 10)  // create 9 additional blocks
```

Because this loop creates 9 blocks (even if the `if` block before created 1).

File: source/agora/utils/Test.d [hunk 12]
```diff
@@ -235,12 +236,10 @@ public Transaction[] makeChainedTransactions (KeyPair key_pair,
 ///
 unittest
 {
-    import agora.consensus.data.Block;
-
     auto gen_key = WK.Keys.Genesis;
 
     /// should spend genesis block's outputs
-    auto txes = makeChainedTransactions(gen_key, null, 1);
+    auto txes = genesisSpendable().map!(txb => txb.sign()).array();
```
Comment (AndrejMitrovic) [inline source/agora/utils/Test.d]: I don't think the changes in these tests make sense. These tests are for the `makeChainedTransactions` function, not for `TxBuilder`.

When a future commit removes `makeChainedTransactions` then these unittests will also be removed. But if `makeChainedTransaction` is not removed yet in this PR then its tests should not be touched.

File: source/agora/utils/Test.d [hunk 13]
```diff
@@ -274,11 +273,9 @@ unittest
 /// example of chaining
 unittest
 {
-    import agora.consensus.data.Block;
-
     auto gen_key = WK.Keys.Genesis;
-    const(Transaction)[] txes = makeChainedTransactions(gen_key, null, 1);
-    txes = makeChainedTransactions(gen_key, txes, 1);
+    auto txs = genesisSpendable().map!(txb => txb.sign()).array();
```
Comment (AndrejMitrovic) [inline source/agora/utils/Test.d]: Ditto about not touching this test.

**Review by AndrejMitrovic — CHANGES_REQUESTED** [review]
A few comments, otherwise looks good.

File: source/agora/test/NetworkManager.d (hunk 9 above)
Comment (kchulj) [reply source/agora/test/NetworkManager.d]: It was not the correct change because it was not creating invalid transactions. I left it untouched in the new commit.
`TxBuilder` does not allow one to build invalid transactions, therefore, it would have to be modified after it has been built.
I will try this in a separate issue.

File: source/agora/common/crypto/Key.d (hunk 2 above)
Comment (kchulj) [reply source/agora/common/crypto/Key.d]: Thanks for the suggestions @AndrejMitrovic 
The problem with the rebase has been fixed and the changes have been applied.
Ready for review again.

**Review by AndrejMitrovic — APPROVED** [review]

File: source/agora/test/BanManager.d [hunk 14]
```diff
@@ -53,15 +53,29 @@ unittest
 
     Transaction[] all_txs;
     Transaction[] last_txs;
+    Transaction[] txes;
```
Comment (Geod24) [inline source/agora/test/BanManager.d]: Why is this outside the function's scope ?

File: source/agora/test/Ledger.d [hunk 15]
```diff
@@ -45,22 +45,25 @@ unittest
 
     Transaction[][] block_txes; /// per-block array of transactions (genesis not included)
     Transaction[] last_txs;
+
+    // get all the blocks (including genesis block)
+    auto blocks = node_1.getBlocksFrom(0, 101);
+    assert(blocks.length == 1);
```
Comment (Geod24) [inline source/agora/test/Ledger.d]: If the expectation is that the length is exactly 1, then:
```suggestion
    // Get the genesis block, make sure it's the only block externalized
    auto blocks = node_1.getBlocksFrom(0, 2);
    assert(blocks.length == 1);
```

File: source/agora/test/NetworkManager.d [hunk 16]
```diff
@@ -23,6 +23,7 @@ import agora.common.Metadata;
 import agora.common.Types;
 import agora.consensus.data.Block;
 import agora.consensus.data.ConsensusParams;
+import agora.consensus.data.genesis.Test;
```
Comment (Geod24) [inline source/agora/test/NetworkManager.d]: Why the import ?

File: source/agora/test/ValidatorCount.d [hunk 17]
```diff
@@ -22,6 +22,7 @@ module agora.test.ValidatorCount;
 
 version (unittest):
 
+import agora.consensus.data.Block;
```
Comment (Geod24) [inline source/agora/test/ValidatorCount.d]: Why the import ?

File: source/agora/test/ValidatorCount.d [hunk 18]
```diff
@@ -45,24 +46,31 @@ unittest
     auto node_1 = nodes[0];
 
     auto gen_key_pair = WK.Keys.Genesis;
+    const(Block)[] blocks = node_1.getBlocksFrom(0, 101);
```
Comment (Geod24) [inline source/agora/test/ValidatorCount.d]: I'll add a comment here / make it consistent with the other call.

**Review by Geod24 — CHANGES_REQUESTED** [review]
Just a few nitpicks, otherwise looks good.

**Review by linked0 — APPROVED** [review]
LGTM

**Review by Geod24 — APPROVED** [review]
